# Проект по SQL 

**Цель исследования**: исследовать базу данных, содержащую информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг для подготовки рекомендации заказчику.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![jupyter](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?table=block&id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=2000&userId=&cache=v2)

**Задания**

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas

In [2]:
#получим первую информацию о таблицах
tab=['books', 'authors', 'publishers', 'ratings', 'reviews']
for table in tab:
    print(f'Таблица {table}')
    query = f'''SELECT * FROM {table} LIMIT 4'''
    con=engine.connect()
    display(pd.io.sql.read_sql(sql=text(query), con = con))

Таблица books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309

Таблица authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert

Таблица publishers


publisher_id      publisher
0             1            Ace
1             2       Ace Book
2             3      Ace Books
3             4  Ace Hardcover

Таблица ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3

Таблица reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...

In [3]:
#получим первую информацию о таблицах
tab=['books', 'authors', 'publishers', 'ratings', 'reviews']
for table in tab:
    print(f'Таблица {table}')
    query = f'''SELECT DISTINCT * FROM {table}'''
    con=engine.connect()
    display(pd.io.sql.read_sql(sql=text(query), con = con))

Таблица books


book_id  author_id                                              title  \
0        640        262                                    The Beach House   
1        174        156                                 Cyrano de Bergerac   
2        392        357                           Like Water for Chocolate   
3        330        449  How to Be a Domestic Goddess: Baking and the A...   
4        448        452              Morrigan's Cross (Circle Trilogy  #1)   
..       ...        ...                                                ...   
995      922        490                                            Thinner   
996      755        599                        The Hunchback of Notre-Dame   
997      758        184                                          The Idiot   
998      384        390                                 Last Chance Saloon   
999       51         54                           A Tree Grows in Brooklyn   

     num_pages publication_date  publisher_id  
0          356       2003-05-01           116  
1          240       2003-08-05           267  
2          222       1993-09-16            51  
3          384       2005-09-01           122  
4          321       2006-08-29           164  
..         ...              ...           ...  
995        320       1996-09-01           265  
996        510       2001-04-10           267  
997        667       2003-04-08           183  
998        528       2003-05-27           135  
999        496       2006-05-30           135  

[1000 rows x 6 columns]

Таблица authors


author_id                             author
0          434            Neil Gaiman/Dave McKean
1          235                      J.D. Salinger
2          451                        Nora Ephron
3          270                    Janet Evanovich
4          238                       J.M. Coetzee
..         ...                                ...
631        159                    Edward P. Jones
632        541                    Stephen Hawking
633         27                      Anna Quindlen
634        339  Kate DiCamillo/Bagram Ibatoulline
635         31                     Anne McCaffrey

[636 rows x 2 columns]

Таблица publishers


publisher_id                         publisher
0             258                          Scribner
1             141                       HarperPrism
2             254                  Scholastic Books
3             142                     HarperTempest
4             295             Top Shelf Productions
..            ...                               ...
335            29                        Avon Books
336             8                      Albin Michel
337            91             Disney Hyperion Books
338           218  Penguin Classics Deluxe Editions
339           207                    Pantheon Books

[340 rows x 2 columns]

Таблица ratings


rating_id  book_id       username  rating
0          6373      978       walter49       4
1           123       28        znelson       4
2          1729      299        rmiller       5
3          5484      848        ewerner       3
4          1541      281      roberto26       4
...         ...      ...            ...     ...
6451        848      144        znelson       5
6452        805      133     kevinquinn       5
6453       5975      928  perrypatricia       2
6454       3554      577  briannachavez       2
6455       1683      299       walter49       5

[6456 rows x 4 columns]

Таблица reviews


review_id  book_id       username  \
0          2046      735  danielpatrick   
1          2738      979  danielpatrick   
2          1220      444    nicoleburns   
3           546      201       yvasquez   
4           238       81  grantpatricia   
...         ...      ...            ...   
2788       1322      479  kevincampbell   
2789        902      334  taylorjeffrey   
2790       2444      875      emcdaniel   
2791       1706      621        susan85   
2792       2711      969        znelson   

                                                   text  
0     Relationship evidence foot quality set. Availa...  
1     Decide record difficult better firm military r...  
2     Pretty under while but method. In total color ...  
3     Police mean ago. View above same fact cold. St...  
4     Cold foreign magazine. Trade cell want factor ...  
...                                                 ...  
2788  Nearly impact according. Someone just treatmen...  
2789  Pass form really. Effect beat staff deal someo...  
2790  Ability be whose believe drop win energy. Say ...  
2791  Fight investment part himself town. Space true...  
2792  Role paper reflect professor feel. Might stage...  

[2793 rows x 4 columns]

**Вывод**

Перед нами 4 таблицы:
- books 1000 записей × 6 столбцов, дубликатов нет
- authors 636 записей × 2 столбца, дубликатов нет
- publishers 340 записи × 2 столбца, дубликатов нет
- ratings 6456 записи × 4 столбца, дубликатов нет
- reviews 2793 записи × 4 столбца, дубликатов нет

## Посчитайте, сколько книг вышло после 1 января 2000 года

In [4]:
print(f'Таблица books')
query = '''
        SELECT COUNT(book_id)
        FROM books
        WHERE publication_date > '2000-01-01'
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Таблица books


count
0    819

**После 1 января 2000 года оппубликованно 819 книг**

In [5]:
print(f'Таблица кол-ва выпущенных книг в разные года')
query = '''
        SELECT EXTRACT(YEAR FROM CAST(publication_date AS timestamp)),
               COUNT(book_id) AS book_cnt
        FROM books
        GROUP BY EXTRACT(YEAR FROM CAST(publication_date AS timestamp))
        ORDER BY book_cnt DESC
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Таблица кол-ва выпущенных книг в разные года


date_part  book_cnt
0      2006.0       184
1      2005.0       139
2      2004.0       124
3      2003.0       105
4      2002.0        94
5      2001.0        60
6      2007.0        50
7      1999.0        41
8      2000.0        38
9      1998.0        30
10     1994.0        14
11     1997.0        14
12     1996.0        13
13     1995.0        11
14     1993.0        10
15     1991.0         9
16     2010.0         7
17     1987.0         7
18     2009.0         6
19     1988.0         5
20     1990.0         5
21     1989.0         5
22     1982.0         5
23     2008.0         4
24     2012.0         3
25     1980.0         2
26     1992.0         2
27     2019.0         2
28     1986.0         2
29     2014.0         2
30     2020.0         1
31     1983.0         1
32     2011.0         1
33     1977.0         1
34     1984.0         1
35     2013.0         1
36     1952.0         1

**Больше всего книг было выпущенно в 2006 году - 184, а всего по 1 книге вышло в разные года:1952, 1977, 1983, 1984, 2011, 2013,2020**

## Для каждой книги посчитайте количество обзоров и среднюю оценку

In [6]:
print(f'Таблица книг с кол-вом обзоров и средней оценкой')
query = '''
        SELECT title,
               COUNT(DISTINCT(review_id)) AS rev_cnt,
               ROUND(AVG(rating), 2) AS avg_rat
        FROM books AS b
        LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
        LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
        GROUP BY b.book_id
        ORDER BY rev_cnt DESC
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Таблица книг с кол-вом обзоров и средней оценкой


title  rev_cnt  avg_rat
0                              Twilight (Twilight  #1)        7     3.66
1                                  Water for Elephants        6     3.98
2                                     The Glass Castle        6     4.21
3    Harry Potter and the Prisoner of Azkaban (Harr...        6     4.41
4    The Curious Incident of the Dog in the Night-Time        6     4.08
..                                                 ...      ...      ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel        0     3.67
996                            The Natural Way to Draw        0     3.00
997   The Cat in the Hat and Other Dr. Seuss Favorites        0     5.00
998                          Essential Tales and Poems        0     4.00
999  Disney's Beauty and the Beast (A Little Golden...        0     4.00

[1000 rows x 3 columns]

**Получилось 1000 книг с рейтингом от 0 до 7, средний балл 3,66. При этом стоит отметить, что кол-во обзоров никак не влияет на рейтинг книг.**

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [7]:
print(f'Издательство, которое выпустило наибольшее число книг толще 50 страниц')
query = '''
        SELECT publisher,
               COUNT(b.book_id) AS books_cnt
        FROM publishers AS p
        LEFT JOIN books AS b ON p.publisher_id=b.publisher_id
        WHERE b.num_pages > 50
        GROUP BY p.publisher
        ORDER BY books_cnt DESC
        LIMIT 1
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Издательство, которое выпустило наибольшее число книг толще 50 страниц


publisher  books_cnt
0  Penguin Books         42

**Издательство Penguin Books выпустило 42 книги обьемом больше 50 страниц. Это британское издательство, основанное в 1935 году в Лондоне. На сегодняшний день Penguin входит в одноименную издательскую группу,  которая в 2011 году стала крупнейшей в мире.**

In [8]:
print(f'Авторы и их книги издательства Penguin Books')
query = '''
        SELECT DISTINCT a.author,
               b.title,
               b.num_pages
        FROM books AS b
        LEFT JOIN authors AS a ON b.author_id=a.author_id
        LEFT JOIN publishers AS p ON b.publisher_id=p.publisher_id
        WHERE p.publisher LIKE '%Penguin Books%'
        GROUP BY a.author, b.title, b.num_pages
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Авторы и их книги издательства Penguin Books


author  \
0                                             Amy Tan   
1                                             Amy Tan   
2                              Anonymous/N.K. Sandars   
3                                      Art Spiegelman   
4                   Arthur  Miller/Christopher Bigsby   
5   Charles Dickens/Nicola Bradbury/Hablot Knight ...   
6                                      David    Allen   
7                                         Don DeLillo   
8       E.M. Forster/Oliver Stallybrass/Pankaj Mishra   
9                    Edgar Rice Burroughs/John Seelye   
10                 Frances Hodgson Burnett/Nancy Bond   
11                                   Geraldine Brooks   
12                                   Gwyn Hyman Rubio   
13   Herodotus/Aubrey de Sélincourt/John M. Marincola   
14                                       J.M. Coetzee   
15                                 Jacquelyn Mitchard   
16                         Jane Austen/Fiona Stafford   
17                                      Jared Diamond   
18                                      Jasper Fforde   
19                                      Jasper Fforde   
20                                      Jasper Fforde   
21                                   Jerome K. Jerome   
22                                       Jodi Picoult   
23                                     John Steinbeck   
24                                     John Steinbeck   
25                                     John Steinbeck   
26                      Kenneth Grahame/Gillian Avery   
27                                        Kim Edwards   
28  Marcus Aurelius/Martin Hammond/Albert Wittstoc...   
29                                      Marisha Pessl   
30                                     Mark Kurlansky   
31                                          Mary Karr   
32  Miguel de Cervantes Saavedra/John Rutherford/R...   
33   Nathaniel Hawthorne/Thomas E. Connolly/Nina Baym   
34                                Nathaniel Philbrick   
35                                        Ron Chernow   
36                      Stephen King/Robin Waterfield   
37                                      Sue Monk Kidd   
38                                      Sue Monk Kidd   
39      Thomas Hardy/Rosemarie Morgan/Shannon Russell   
40                                     Thomas Pynchon   
41                                     Virginia Woolf   
42                                   Vladimir Nabokov   
43                        Walter Scott/Graham Tulloch   
44                                    William Golding   
45             William Makepeace Thackeray/John Carey   
46                                        Zadie Smith   

                                                title  num_pages  
0                                   The Joy Luck Club        288  
1                              The Kitchen God's Wife        416  
2                               The Epic of Gilgamesh        120  
3                                   The Complete Maus        296  
4                                        The Crucible        143  
5                                         Bleak House       1017  
6   Getting Things Done: The Art of Stress-Free Pr...        267  
7                                         White Noise        320  
8                                  A Passage to India        376  
9                    A Princess of Mars (Barsoom  #1)        186  
10                                  A Little Princess        242  
11                                    Year of Wonders        358  
12                                         Icy Sparks        320  
13                                      The Histories        716  
14                                           Disgrace        220  
15   The Deep End of the Ocean (Cappadora Family  #1)        464  
16                                               Emma        474  
17  Collapse: How Societies Choose to Fail or Succeed        608  
18            Lost in a Good Book (Th

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [9]:
print(f'Автор с самой высокой средней оценкой книг')
query = '''
        WITH
        i AS (SELECT r.book_id,
              ROUND(AVG(rating), 2) AS avg_rating
              FROM ratings AS r
              GROUP BY r.book_id
              HAVING COUNT(r.book_id) >= 50)
                              
        SELECT a.author,
               avg_rating
        FROM i
        LEFT JOIN books AS b ON i.book_id=b.book_id
        LEFT JOIN authors AS a ON b.author_id=a.author_id
        GROUP BY a.author, avg_rating
        LIMIT 1
        '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Автор с самой высокой средней оценкой книг


author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

**Автор с самой высокой средней оценкой книг - стала J.K. Rowling/Mary GrandPré - 4,29 средняя оценка. Кстати, Mary GrandPré этот иллюстратор которая как раз работала над первыми книгами о Гарри Поттере.**

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [10]:
print(f'Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок')
query = '''     
        SELECT ROUND(AVG(cnt_reviews),2) AS avg_cnt_review
        FROM (
              SELECT rev.username,
              COUNT(rev.review_id) AS cnt_reviews
              FROM reviews AS rev
              WHERE rev.username IN (SELECT rat.username
              FROM ratings AS rat
              GROUP BY rat.username
              HAVING COUNT(rat.rating_id) > 48)
              GROUP BY rev.username
             ) AS cnt_rev
        '''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок


avg_cnt_review
0            24.0

**Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 обзора**

<div style="border:solid green 2px; padding: 20px">
<b>Вывод</b>
    
Цель исследования была проанализировать базу данных, содержащую информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг для подготовки рекомендации заказчику.   

Были получены следующие ответы на поставление задачи:
    
- Посчитайте, сколько книг вышло после 1 января 2000 года - 819
      
- Для каждой книги посчитайте количество обзоров и среднюю оценку - Получилось 1000 книг с рейтингом от 0 до 7, средний балл 3,66.
    
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц -Издательство Penguin Books выпустило 42 книги
    
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками - стала J.K. Rowling/Mary GrandPré - 4,29
    
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 обзора.
    
    
    Больше всего книг было выпущенно в 2006 году - 184, а всего по 1 книге вышло в разные года:1952, 1977, 1983, 1984, 2011, 2013,2020. Подробнее рассмотрели авторов издающихся в издательстве Penguin Books. 
    
**Рекомендации**
    
- Стоит обратить на формат аудиокниг, он тоже достаточно популярен. 
- За кол-во обзоров и рецензии давать какие-нибудь "плюшки" - купоны, закрытые акции, мерч.
- За количетсво купленых книг присуждать статусы или элементы самовыражения
- Устраивать творческие конкурсы и т.п